pyg_lib/csrc/random/cpu/biased_sampling.h

```diff
@@ -1,21 +1,134 @@
 #pragma once
 
+#include <algorithm>
+#include <iterator>
+
+#include "pyg_lib/csrc/random/cpu/rand_engine.h"
+
 namespace pyg {
 namespace random {
 
-// TODO: Biased random sampling methods
+// The two following functions randomly choose one index according to the given
+// bias array.
+// The probability for an index to be chosen is proportional to the
+// corresponding bias.
+
+/**
+ * Biased random choice from a preprocessed CDF (exclusive sum) bias array
+ *
+ * @tparam Idx type of the set of indices to choose from
+ *
+ * @tparam Bias type of the weight array
+ *
+ * @param idx the pointer to the start of the index array
+ *
+ * @param bias the pointer to the start of the bias array
+ *
+ * @param len the length of the index/bias arraay
+ *
+ * @returns the chosen index
+ *
+ * Use binary search to map a 0-1 uniform random real number X to idx[i]
+ * s.t. idx[i] < X < idx[i+1]
+ *
+ * Example:
+ * idx: { 0,   1,  |  2 }
+ * cdf: {0.0, 0.2, | 0.8}
+ *                 |
+ *              X:0.7
+ * -> choose idx = 1
+ */
 
 template <typename Idx, typename Bias>
```

Comment:
```suggestion
template <typename index_t, typename scalar_t>
```
to align with naming of other files?